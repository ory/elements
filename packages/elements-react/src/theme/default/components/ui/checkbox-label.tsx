// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import { UiText } from "@ory/client-fetch"
import { useIntl } from "react-intl"
import { uiTextToFormattedMessage } from "../../../../util"

type CheckboxLabelProps = {
  label?: UiText
}

const linkRegex = /\[([^\]]+)\]\(([^)]+)\)/g

export function computeLabelElements(labelText: string) {
  const elements = []
  let lastIndex = 0

  // Use matchAll to find all markdown links
  for (const match of labelText.matchAll(linkRegex)) {
    const linkText = match[1]
    const url = match[2]
    const matchStart = match.index
    if (typeof matchStart === "undefined") {
      // Some types seem to be wrong somewhere, eslint complains that matchStart can be undefined, but it can't?
      // So we just skip this match, if it is undefined
      continue
    }

    // Push the text before the match
    if (matchStart > lastIndex) {
      elements.push(labelText.slice(lastIndex, matchStart))
    }

    // Push the <a> tag for the markdown link
    elements.push(
      <a
        key={matchStart}
        href={url}
        target="_blank"
        rel="noopener noreferrer"
        className="text-button-link-brand-brand hover:text-button-link-brand-brand-hover underline"
      >
        {linkText}
      </a>,
    )

    // Update lastIndex to the end of the current match
    lastIndex = matchStart + match[0].length
  }

  // Push any remaining text after the last match
  if (lastIndex < labelText.length) {
    elements.push(labelText.slice(lastIndex))
  }
  return elements
}

export function CheckboxLabel({ label }: CheckboxLabelProps) {
  const intl = useIntl()
  if (!label) {
    return null
  }

  const labelText = uiTextToFormattedMessage(label, intl)

  return <>{computeLabelElements(labelText)}</>
}
