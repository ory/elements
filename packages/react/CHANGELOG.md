# Change Log

All notable changes to this project will be documented in this file. See
[Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.0.1-alpha.1](https://github.com/ory/elements/compare/v0.0.1-alpha.0...v0.0.1-alpha.1) (2022-09-27)

**Note:** Version bump only for package @ory/elements
