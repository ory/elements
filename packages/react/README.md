# Ory Elements

> [!CAUTION] This package is in maintenance mode and is no longer actively
> developed. Please use [`@ory/elements-react`](../elements-react/README.md)
> instead.

Ory Elements is a component library that makes building login, registration and
account pages for Ory a breeze.

- Reduces time to add complex auth flows to your customer experience, including
  multi-factor authentication and account recovery
- Themeable and modular - use only what you need from it
- Works with the React ecosystem (NextJS, React SPA, Preact SPA)
- Works with the Express based ecosystem
- Dynamically adapts the user interface to your Ory identity schema, sign-in and
  flow configuration

Ory Elements supports integrating with:

- React
- Preact
- ExpressJs (experimental)

## Getting Started

Install Ory Elements into your React application

```shell
npm i @ory/elements --save
```

Ory Elements provides you with the ability to theme your application. Wrap your
components with `<ThemeProvider />` and import `style.css` from Ory Elements.

It is best to add this to your `main.tsx` file since it will be reused across
your project. All source code is available in the Ory Elements
[examples/react-spa](https://github.com/ory/elements/tree/main/examples/react-spa)
folder.

```tsx
// other imports here
// Ory Elements
// optional global css reset
import "@ory/elements/assets/normalize.css"
// optional fontawesome icons
import "@ory/elements/assets/fa-brands.min.css"
import "@ory/elements/assets/fa-solid.min.css"
import "@ory/elements/assets/fontawesome.min.css"

// optional fonts
import "@ory/elements/assets/inter-font.css"
import "@ory/elements/assets/jetbrains-mono-font.css"

// required styles for Ory Elements
import "@ory/elements/style.css"

ReactDOM.createRoot(document.getElementById("root") as HTMLElement).render(
  <React.StrictMode>
    <BrowserRouter>
      {/* We add the Ory themes here */}
      <ThemeProvider themeOverrides={customTheme}>
        <Routes>
          <Route path="/" element={<Dashboard />} />
          <Route path="/login" element={<Login />} />
          <Route path="/signup" element={<Registration />} />
          <Route path="/verification" element={<Verification />} />
          <Route path="/recovery" element={<Recovery />} />
          <Route path="/settings" element={<Settings />} />
        </Routes>
      </ThemeProvider>
    </BrowserRouter>
  </React.StrictMode>,
)
```

Ory Elements supports integrating with Ory out of the box with the
`<UserAuthCard/>`, `<UserSettingsCard />` and `<UserErrorCard />`.

Each flow can be mapped using these components with only a few lines of code.

Let's create a login example, create `login.tsx`.

```tsx
exoort const Login = () => {
  // do sdk calls to get the flow data from Ory
  return flow ? (
    <UserAuthCard
      flow={flow}
      flowType={"login"}
      additionalProps={{
        forgotPasswordURL: "/recovery",
        signupURL: "/signup",
      }}
      title={"Login"}
      includeScripts={true}
      onSubmit={submitFormData}
    />
  ) : (
    <div>Loading...</div>
  )
}
```
