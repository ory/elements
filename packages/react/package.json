{
  "name": "@ory/elements",
  "version": "0.0.0",
  "publishConfig": {
    "access": "public"
  },
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs",
      "require": "./dist/index.umd.js"
    },
    "./style.css": "./dist/style.css",
    "./package.json": "./package.json",
    "./assets/*": "./dist/assets/*",
    "./assets/**/*": "./dist/assets/**/*",
    "./assets/fa-brands.min.css": "./dist/assets/fa-brands.min.css",
    "./assets/fa-solid.min.css": "./dist/assets/fa-solid.min.css",
    "./assets/fontawesome.min.css": "./dist/assets/fontawesome.min.css",
    "./assets/inter-font.css": "./dist/assets/inter-font.css",
    "./assets/jetbrains-mono-font.css": "./dist/assets/jetbrains-mono-font.css",
    "./assets/normalize.css": "./dist/assets/normalize.css",
    "./assets/webfonts/fa-brands-400.ttf": "./dist/assets/webfonts/fa-brands-400.ttf",
    "./assets/webfonts/fa-brands-400.woff2": "./dist/assets/webfonts/fa-brands-400.woff2",
    "./assets/webfonts/fa-solid-900.ttf": "./dist/assets/webfonts/fa-solid-900.ttf",
    "./assets/webfonts/fa-solid-900.woff2": "./dist/assets/webfonts/fa-solid-900.woff2",
    "./assets/jetbrains-mono/JetBrainsMono-Bold.woff2": "./dist/assets/jetbrains-mono/JetBrainsMono-Bold.woff2",
    "./assets/jetbrains-mono/JetBrainsMono-Regular.woff2": "./dist/assets/jetbrains-mono/JetBrainsMono-Regular.woff2",
    "./assets/inter/Inter-Bold.woff": "./dist/assets/inter/Inter-Bold.woff",
    "./assets/inter/Inter-Bold.woff2": "./dist/assets/inter/Inter-Bold.woff2",
    "./assets/inter/Inter-Regular.woff": "./dist/assets/inter/Inter-Regular.woff",
    "./assets/inter/Inter-Regular.woff2": "./dist/assets/inter/Inter-Regular.woff2",
    "./assets/inter/Inter-SemiBold.woff": "./dist/assets/inter/Inter-SemiBold.woff",
    "./assets/inter/Inter-SemiBold.woff2": "./dist/assets/inter/Inter-SemiBold.woff2"
  },
  "main": "./dist/index.umd.js",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "scripts": {
    "build": "vite build"
  },
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com/ory/elements.git",
    "directory": "packages/react"
  },
  "license": "Apache License 2.0",
  "bugs": {
    "url": "https://github.com/ory/elements/issues"
  },
  "homepage": "https://github.com/ory/elements#readme",
  "files": [
    "dist/*"
  ],
  "engines": {
    "node": ">=16.16.0",
    "npm": ">=8.11.0"
  },
  "typedoc": {
    "entryPoint": "../../src/react.ts"
  }
}
