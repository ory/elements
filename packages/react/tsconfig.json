{
  "compilerOptions": {
    "rootDir": "../../src",
    "jsx": "preserve"
  },
  "extends": "../../tsconfig.build.json",
  "include": ["../../src", "vite.config.mts", "package.json"]
}
