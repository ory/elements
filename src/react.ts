// Copyright Â© 2023 Ory Corp
// SPDX-License-Identifier: Apache-2.0

export * from "./common"
export * from "./react-components"
export * from "./react-components/ory/helpers/types"
export * from "./theme"
export * as locales from "./locales"
