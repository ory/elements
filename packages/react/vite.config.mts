// Copyright Â© 2023 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import { vanillaExtractPlugin } from "@vanilla-extract/vite-plugin"
import react from "@vitejs/plugin-react"
import path from "path"
import { defineConfig } from "vite"
import dts from "vite-plugin-dts"
import { viteStaticCopy } from "vite-plugin-static-copy"
import autoprefixer from "autoprefixer"

// https://vitejs.dev/config/
export default defineConfig({
  css: {
    postcss: {
      plugins: [
        autoprefixer({}), // add options if needed
      ],
    },
  },
  plugins: [
    vanillaExtractPlugin({
      identifiers: ({ hash, filePath, debugId }) => {
        const name = filePath
          .split("/")
          ?.pop()
          ?.split(".")[0]
          ?.replace("-", "_")
        const id = debugId ? "_" + debugId : ""
        return `ory_elements__${name}${id}__${hash}`
      },
    }),
    dts({
      insertTypesEntry: true,
    }),
    react(),
    viteStaticCopy({
      targets: [
        {
          src: "../../src/assets",
          dest: "",
        },
      ],
    }),
  ],
  build: {
    target: "esnext",
    lib: {
      name: "@ory/elements",
      entry: path.resolve(__dirname, "../../src/react.ts"),
      formats: ["es", "umd"],
      fileName: (format) => (format === "es" ? "index.mjs" : "index.umd.js"),
    },
    rollupOptions: {
      treeshake: "smallest",
      external: ["react", "react-dom"],
    },
  },
})
