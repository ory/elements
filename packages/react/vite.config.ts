import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';
import { vanillaExtractPlugin } from '@vanilla-extract/vite-plugin';
import dts from 'vite-plugin-dts';
import path from 'path';

// https://vitejs.dev/config/
export default defineConfig({
  /* WARNING: Vanilla Extract does not support the latest Vite (3.x) version yet.*/
    esbuild: {
      minify: true
    },
    plugins: [
      vanillaExtractPlugin(),
      dts({
        insertTypesEntry: true,
      }),
      react()
    ],
    build: {
      target: 'esnext',
      lib: {
        name: '@ory/themes',
        entry: path.resolve(__dirname, '../../src/react.ts'),
        formats: ['es', 'umd'],
        fileName: (format) => (format === 'es' ? 'index.mjs' : 'index.umd.js')
      }
      // rollupOptions: {
      //   treeshake: 'smallest',
      //   external: ['react', 'react-dom']
      //   outputs: {
      //     global: {
      //       react: 'React',
      //       'react-dom': 'ReactDOM',
      //       'react-dom/server': 'ReactDOMServer'
      //     }
      //   }
      // }
    }
  });
