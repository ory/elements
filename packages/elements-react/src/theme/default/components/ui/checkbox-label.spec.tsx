// Copyright Â© 2024 Ory Corp
// SPDX-License-Identifier: Apache-2.0

import { render } from "@testing-library/react"
import { CheckboxLabel } from "./checkbox-label"
import { IntlProvider } from "react-intl"
import { PropsWithChildren } from "react"

const wrapper = ({ children }: PropsWithChildren) => (
  <IntlProvider locale="en">{children}</IntlProvider>
)

describe("computeLabelElements", () => {
  test("renders plain text without links correctly", () => {
    const labelText = "This is just plain text"

    const { container } = render(
      <CheckboxLabel label={{ text: labelText, id: 0, type: "info" }} />,
      { wrapper },
    )
    expect(container).toMatchSnapshot()
  })

  test("renders a text with a single markdown link correctly", () => {
    const labelText = "This is a [link](https://example.com)"

    const { container } = render(
      <CheckboxLabel label={{ text: labelText, id: 0, type: "info" }} />,
      { wrapper },
    )
    expect(container).toMatchSnapshot()
  })

  test("renders a text with multiple markdown links correctly", () => {
    const labelText =
      "This [first link](https://first.com) and this [second link](https://second.com)"

    const { container } = render(
      <CheckboxLabel label={{ text: labelText, id: 0, type: "info" }} />,
      { wrapper },
    )
    expect(container).toMatchSnapshot()
  })

  test("renders a text with link and extra text around it correctly", () => {
    const labelText =
      "Click [here](https://example.com) to visit, or go elsewhere."

    const { container } = render(
      <CheckboxLabel label={{ text: labelText, id: 0, type: "info" }} />,
      { wrapper },
    )
    expect(container).toMatchSnapshot()
  })

  test("handles a label with no text but a link", () => {
    const labelText = "[Click here](https://example.com)"

    const { container } = render(
      <CheckboxLabel label={{ text: labelText, id: 0, type: "info" }} />,
      { wrapper },
    )
    expect(container).toMatchSnapshot()
  })

  test("renders null if label is undefined", () => {
    const { container } = render(<CheckboxLabel label={undefined} />, {
      wrapper,
    })
    expect(container).toMatchSnapshot()
  })
})
