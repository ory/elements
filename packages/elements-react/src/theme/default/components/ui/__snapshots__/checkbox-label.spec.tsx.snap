// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computeLabelElements handles a label with no text but a link 1`] = `
<div>
  <a
    class="text-button-link-brand-brand hover:text-button-link-brand-brand-hover underline"
    href="https://example.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    Click here
  </a>
</div>
`;

exports[`computeLabelElements renders a text with a single markdown link correctly 1`] = `
<div>
  This is a 
  <a
    class="text-button-link-brand-brand hover:text-button-link-brand-brand-hover underline"
    href="https://example.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    link
  </a>
</div>
`;

exports[`computeLabelElements renders a text with link and extra text around it correctly 1`] = `
<div>
  Click 
  <a
    class="text-button-link-brand-brand hover:text-button-link-brand-brand-hover underline"
    href="https://example.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    here
  </a>
   to visit, or go elsewhere.
</div>
`;

exports[`computeLabelElements renders a text with multiple markdown links correctly 1`] = `
<div>
  This 
  <a
    class="text-button-link-brand-brand hover:text-button-link-brand-brand-hover underline"
    href="https://first.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    first link
  </a>
   and this 
  <a
    class="text-button-link-brand-brand hover:text-button-link-brand-brand-hover underline"
    href="https://second.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    second link
  </a>
</div>
`;

exports[`computeLabelElements renders null if label is undefined 1`] = `<div />`;

exports[`computeLabelElements renders plain text without links correctly 1`] = `
<div>
  This is just plain text
</div>
`;
